Implement Shooter

aidnem opened the pull request:
Adds:

- Scoring subsystem docs under docs/subsystems/scoring (for scoring subsystem and shooter)
- Adds test mode features (test mode manager, real testPeriodic in RobotContainer)
- CANConstants for managing CAN bus IDs (hopefully this will help identify conflicts more easily than having CAN IDs spread across the project)

## Verification

### Closed-loop works in sim:

<img width="1131" height="1001" alt="image" src="https://github.com/user-attachments/assets/7cb31c55-1627-4aa7-ab8e-590dc4f375a8" />

Left axis:
- Left target RPM (blue)
- Left actual RPM (orange)
Right axis:
- Right target RPM (green)
- Right actual RPM (red)

You can observe each of the actual values following their targets quite well.

### Pose-based shot calculations work in sim:

https://github.com/user-attachments/assets/0038ac56-96f6-4ece-b75d-8c927d7c4f0f

The blue line is the unclamped target shot (The shot that's directly projected out of the side of the robot). The green line is the shot that it is actually warming up for, which is the shot but clamped to be within the bounds of the barge. This was tested on blue and red and it worked on both.

You can observe the clamped shot distance on the left stays between 2 and 10 (the min and max with the mock numbers I put in for testing). You can also see that the leftRPM and rightRPM are appropriate compared to each other (leftRPM < rightRPM, since I put in lower numbers for left than right RPM in the shooter map).

This is now ready for review ✅

aidnem committed:
(#8) Move shooter to its own folder and add ShooterIO

aidnem committed:
(#5) WIP - Start adding shooter logic

aidnem committed:
(#8) Use JsonConstants for ShooterConstants loading and implement some Shooter stubs

aidnem committed:
(#8) Add basic docs on scoring subsystem and shooter mechanism

aidnem committed:
(#8) Try to fix links in scoring docs

aidnem committed:
(#8) Add ShooterIOTalonFX

aidnem committed:
(#8) WIP (untested) - Add ShooterIOSim

aidnem committed:
(#8) WIP - Sim not functional - Add Test mode

aidnem committed:
(#8) Enable continuous closed-loop and apply gearing calculation, sim still doesn't work

aidnem committed:
(#8) Potentially fix bug from reusing the same cached unit object in ShooterIOSim

aidnem committed:
(#8) Make gradlew executable

aidnem committed:
(#8) Try negating motor inputs in Sim IO, fixes DiscordantMotoring bug but now current draw is ridiculously high

aidnem committed:
(#8) Refactor shooterIO to use 1 for each side, removing a ton of duplicated code

401User2025 committed:
(#8) Add comment about logging

aidnem committed:
(#8) WIP - Simplify sim code. It now works but only CCW with gearing=1.0

aidnem committed:
(#8) WIP - Shooter Sim now works with both motor inverts, gearing is still untested

aidnem committed:
(#8) Add max acceleration tuning to shooter, tune shooter. Shooter sim now functions with closed-loop

aidnem commented:
Currently all of the code is in terms of motor RPM and not flywheel RPM. I think this is okay because, since we aren't doing any physics equations for finding distances (I asked my AP Physics teacher and he said it would be pretty implausible to model this problem using physics) we shouldn't really need the shooter RPM for any reason. If this is objectionable, I can add the flywheel-speed-based setpointing capability pretty easily.

aidnem committed:
(#8) Add closed loop reference logging to ShooterIOTalonFX

aidnem commented:
The only work remaining now is to add proper ShooterSpeeds logging and add the speed calculations from poses. Both are in the works for tomorrow. Sorry for the premature review request, I went back and checked the issue later and realized I wasn't done.

aidnem committed:
(#8) Make ShooterSpeeds into a StructSerializable class so it can be logged by akit

aidnem committed:
(#8) Add pose-based shots to shooter

aidnem committed:
(#8) Update shooter docs to reflect changed API and sync all JSON constants

aidnem committed:
(#8) Format JSON

aidnem commented:
@godmar all functionality has now been tested, so I'm ready for a review whenever you have time. Thanks!

godmar commented on docs/subsystems/scoring/shooter.adoc:
```diff
@@ -0,0 +1,42 @@
+
+=== Shooter closed loop tuning
+
+This mode will read PID & FeedForward gains from NetworkTables and use them to control to a setpoint specified for each wheel goal speed (also read from NetworkTables).
```
Is this the "wheel speed" after transmission or the "motor" goal speed?

aidnem replied:
This should actually read "motor" goal speed.

godmar commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -77,9 +80,17 @@ private void loadConstants() {
+      scoring = InitSubsystems.initScoring();
+
+      if (JsonConstants.featureFlags.runDrive) {
+        scoring.initializeShooterPoseSupplier(() -> drive.getPose());
```
I assume it can be assumed that `drive != null` here (?)

aidnem replied:
Yes. I've added a comment to clarify.

godmar commented on src/main/java/frc/robot/constants/BlueFieldLocations.java:
```diff
@@ -0,0 +1,29 @@
+   */
+  public final Pair<Translation2d, Translation2d> bargeLine =
+      new Pair<Translation2d, Translation2d>(
+          new Translation2d(8.758, 4.566), new Translation2d(8.758, 7.760));
```
My first instinct here is yuk, where do these constants from. I assume they're in meters relative to the field coordinate system?

aidnem replied:
Yes. I'll add a comment to explain the origin of these values.

godmar commented on src/main/java/frc/robot/constants/CANConstants.java:
```diff
@@ -0,0 +1,29 @@
+          EnvironmentHandler.getEnvironmentHandler().getEnvironmentPathProvider(),
+          new JSONSyncConfigBuilder().setPrettyPrinting(true).build());
+
+  public final Integer shooterLeftMotorID = 13; // TODO: Real motor IDs
```
the TODO can be removed.
Also, just a side note - these are default values here, I suppose. In actual operations an actual configuration is being read. Is it better to set them to `null` then to detect when the actual configuration isn't being read, or is it better to have some likely correct values here?

aidnem replied:
In my opinion the TODO should remain until we build a canonical CAN ID document. I think keeping those IDs up to date in CANConstants.java is the way this process should take place (last year we also used a Google Sheet, but I like keeping the constants in code). However, until we flash the actual motors on the robot with this ID, it remains only a preference for what the ID will eventually be.

The way we've always done it was to have likely correct values here. We would update the constants in JSON in a time-sensitive context (e.g. when changing values during shop time when compile time put a significant damper on progress) but would periodically update the java constants to reflect the JSON changes so that we had 2 redundant copies of the correct constants (which would usually take place when we merged a large PR). My preference is to continue this system, because then if something were to happen to the JSON files (e.g. they were mistakenly deleted or some field were missing) the Java file would still reflect the correct value in most cases. I'm very open to discuss this, please let me know what you think.

godmar replied:
Well, as with any redundancy system, there's no problem if the values match. The question is what happens when you have divergent values in multiple places.

aidnem replied:
JSON values will take precedence and replace the Java values.

godmar commented on src/main/java/frc/robot/constants/subsystems/scoring/ShooterConstants.java:
```diff
@@ -1,5 +1,190 @@
+          .withClosedLoopGeneral(new ClosedLoopGeneralConfigs().withContinuousWrap(true))
+          .withSlot0(
+              new Slot0Configs()
+                  .withKP(100.0) // TODO: Tune gains in real life
```
Why aren't these constants coming from JSON?

aidnem replied:
They are, but I wrote the config object in Java first before serializing it. It needs the field to exist so it has something to overwrite when it loads JSON I think.

@avidraccoon can/should these be null in code once the JSON constants exist? I personally lean toward leaving sane defaults present in the Java files but we can evaluate more.

avidraccoon replied:
I think they can be removed, but I agree they should just be left in the Java files.

godmar replied:
I was asking about the `100.0` constant literal - what's here is the fall back in case `.withKP()` isn't later called with the correct constant?

aidnem replied:
This whole object will be getting replaced from JSON, so you don't need `.withKP()` to be called later. The `100.0` constant literal should be the correct constant 80% of the time since I'm updating the gains from the JSON to the Java fairly often. However, the object from JSON will fully replace this object as far as I'm aware. Therefore, the literal should have no effect as long as the field exists in JSON.

@avidraccoon `synced.loadData()` creates an object and THEN replaces all the fields, right? This means that all of this instance-level code will be overridden anyway, as far as I understand.

godmar commented on src/main/java/frc/robot/constants/subsystems/scoring/ShooterConstants.java:
```diff
@@ -1,5 +1,190 @@
+      new ShooterSpeeds(RPM.of(100), RPM.of(50)); // TODO: Tune this in real life!
+
+  /** The set of distances for the mapping of distance to shooter speeds */
+  public final double[] shooterMapDistances = {2.0, 10.0};
```
in meters, I assume?

godmar commented on src/main/java/frc/robot/constants/subsystems/scoring/ShooterConstants.java:
```diff
@@ -1,5 +1,190 @@
+   * <p>This method MUST be called AFTER `synced.loadData()` but BEFORE the maps or max distance are
+   * read.
+   */
+  public void initializeSpeedMaps() {
```
or `initializeDistanceToRPMMaps` since that's what they're called above?

aidnem replied:
That is a better name, thank you

godmar commented on src/main/java/frc/robot/constants/subsystems/scoring/ShooterConstants.java:
```diff
@@ -1,5 +1,190 @@
+  /**
+   * The furthest distance in the shooter map.
+   *
+   * <p>`initializeSpeedMaps()` MUST be called AFTER `synced.loadData()` but BEFORE this value is
```
Side note: a future project might be to extend JSONSync to call some methods annotated with @AfterJSONLoad that could automatically run after JSON has been loaded and initialize such dependent values.

aidnem replied:
Good idea! I created [this issue](https://github.com/team401/coppercore/issues/163) to track this idea.

godmar commented on src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java:
```diff
@@ -48,6 +51,16 @@ public static ScoringSubsystem create(IndexerMechanism indexer, ShooterMechanism
   }
 
+  /**
+   * Initalize the Shooter mechanism's drive pose supplier to use for pose-based shots
```
typo

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterIO.java:
```diff
@@ -0,0 +1,106 @@
+  public default void stop() {}
+
+  /**
+   * Run the shooter flywheel at a certain speed using Motion Magic Velocity (TorqueCurrentFOC)
```
This is motor speed or flywheel speed?

aidnem replied:
This is motor speed. All speeds are in terms of the motor for now, until we find a reason to use flywheel speeds instead. Since we aren't using physics equations for anything, I believe that motor speeds are just a simpler option and won't ruin any math.

godmar replied:
might be worth noting in the code also and not just in the github issue

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterIOSim.java:
```diff
@@ -0,0 +1,71 @@
+    // Flywheel Speed = Motor Speed * Gearing, therefore:
+    // Motor Speed = Flywheel Speed / Gearing
+    // The same math holds for position
+    motorSimState.addRotorPosition(
```
What kind of transmission is being used here? If the motor and the fly wheel rotate in opposite directions, would you need to multiply the velocity with negative 1 * gearing here to obtain the correct rotor position?

aidnem replied:
That was my thought too, but when I multiplied the velocity by -1 the sim behaved as if the motor was being back driven (as far as I could tell). I believe this is because the chassis orientation in the Sim state is a separate invert value from the invert that's present in the config.

I can look at CAD to see the direction the motor is actually spinning but I think it works as is.

godmar replied:
The transmission is a belt drive that doesn't change the direction of rotation, is that true?

aidnem replied:
It's gears, so the direction is technically inverted. However, the motor sim doesn't know that it's inverted and since the velocity is being commanded and read from the motor's POV, I believe all that needs to happen is that the motor should always spin "forward" (positive) in its view. We'll use motor inverts to determine which way each flywheel turns on the actual robot.

Here's a screenshot of CAD:

<img width="1238" height="665" alt="Image" src="https://github.com/user-attachments/assets/de12800f-d0fd-4c1e-a212-df98b07fc10c" />

aidnem replied:
Upon further consideration, I've realized that it doesn't really matter which way the flywheel spins compared to the motor in simulation. As long as the motor, when power is applied, turns in the forward direction (as configured in the motor invert constants) and reaches its target speed, it will work in real life, since we can just flip the motor invert if the flywheel spins backward. @godmar do you agree?

godmar replied:
I agree, in principle, given how easy it is to invert motors. Still haven't had time to fully get the simulation to run :-(

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterIOTalonFX.java:
```diff
@@ -0,0 +1,200 @@
+    ParentDevice.optimizeBusUtilizationForAll(motor);
+  }
+
+  /** Apply the current talonFXConfigs to the motor, trying to re-apply until it succeeds */
```
one stands for wonder what happens if it doesn't succeed after maxConfigApplyAttempts. Presumably, this would be so catastrophic we wouldn't have to worry about because there's no way to recover?

aidnem replied:
I'm unsure. I believe the thinking is that we limit the number of config applies to prevent code from hanging indefinitely on startup. I'm not entirely sure of whether the configs persist after a power cycle on the motors, which, if true, would mean that it would not be catastrophic if we failed to apply maxConfigAttempts times. @PChild can you confirm that motor configs will remain correct on the motors after they've been flashed once (although it isn't guaranteed, for instance in the case of motors being replaced or configs being reset through Tuner X)?

I will add a log message so that we can know if this ever happens. I'll also see if I can make it send a [WPILib Alert](https://docs.wpilib.org/en/stable/docs/software/telemetry/persistent-alerts.html) so we can see it in Elastic dashboard if the failure ever occurs.

PChild replied:
Ideally the controllers are configured correctly during system bring up and the configs are set in controller memory and then these reconfigures don't really do much, they're just a safeguard against if we do replace a motor or it somehow loses its config (which does unfortunately happen sometimes). We definitely don't want to assume that the correct config is applied and don't want to block until we absolutely successfully apply the config, I think this is a reasonable strategy for now.

godmar commented on src/main/java/frc/robot/util/GeomUtil.java:
```diff
@@ -0,0 +1,21 @@
+   *
+   * @param a The first translation (a), representing a 2d vector
+   * @param b The second translation (b), representing a 2d vector
+   * @return If the vectors are collinear, 0. If b is clockwise from a, a negative value. If b is
```
I don't see where this method is being used in this branch, but don't forget that if you use `double` you generally will not get exactly 0. For instance, `0.1 * 308.7 - 44.1 * 0.7` is equal to `3.552713678800501e-15` under IEEE-754 double arithmetic. What to do depends on the use case.

aidnem replied:
Updated the javadoc comment to be more specific. Thanks for catching that!
This is used in `ShooterMechanism::calculatePoseBasedSpeeds()` (line 478, 493) for some of the intersection calculations.

godmar replied:
Weird. I didn't see `calculatePoseBasedSpeeds` at all in this review.
Oh, I see why. github hides large diffs by default.

aidnem committed:
(#8) Clarify shooter constant names (review requested)

aidnem committed:
(#8) Make shooter distance->RPM maps automatically determine minShotDistance based on shortest distance in the map

aidnem committed:
(#8) Fix review comments: clarify names, doc comments, method behavior.

- Clarify units in names of Shooter Constants
- Fix typo in ShooterMechanism.java
- Add javadocs to FieldLocations constant files
- Clarify behavior in GeomUtil.cross doc comment
- Improve shooter constants initialization method name

aidnem committed:
(#8) Make tryUntilOk return whether or not it succeeded and make ShooterIOTalonFX print an error & display an alert when failing to apply a config

aidnem committed:
(#8) Update shooter adoc to reflect that goal speeds are in terms of the motor

aidnem committed:
(#8) Clarify ShooterIO.runSpeed doc comment to indicate that it is in terms of motor speed

aidnem committed:
(#8) Revert deletion of PathPlanner folder because navgrid and settings may be used later for OTF pathing

aidnem commented:
This should be ready to review again, I believe I've cleaned up all that I needed to.

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterMechanism.java:
```diff
@@ -0,0 +1,595 @@
+      }
+      case WARMUP -> {
+        if (poseSupplierInitialized && poseBasedShooting) {
+          ShooterSpeeds speeds = calculatePoseBasedSpeeds();
```
I have a curiosity question here. Why is this run in every periodic rather than only if the pose supplier provides a new/different value?

aidnem replied:
@godmar my thinking was that the performance impact wasn't significant enough to warrant checking if we'd moved before calculating, but I could add a way to check if we're within a certain epsilon of our previous position if the cycle times are bad enough. I suppose it would technically save some battery if the code did less calculations even if it turned out that our cycle times aren't impacted.

godmar replied:
I agree with the sentiment of benchmarking cycle times first.

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterMechanism.java:
```diff
@@ -0,0 +1,595 @@
+    // Keep track of whether we're shooting left or right
+    final boolean isShotLeft;
+
+    if (AllianceUtil.isRed()) {
```
you could shorten this by setting `isShotleft` first and then computing shotDirection and recordOutput only once.

aidnem replied:
Good point. Since we can only shoot left now, I'm removing it entirely.

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterMechanism.java:
```diff
@@ -0,0 +1,595 @@
+
+    Notes:
+        - If s × d = 0, the segment and ray are parallel.
+        - Valid intersection requires u ≥ 0 (in front of ray).
```
I don't see where you compute and check u (?)

godmar replied:
You added this below.

aidnem committed:
(#8) Make shooter only able to shoot left, add InitBindings file

aidnem commented:
@godmar in light of Wednesday's prototyping, I've updated the code so that the robot can only shoot left. I also took the liberty of improving the binding configuration so that hopefully it will be better structured.

aidnem committed:
(#8) Add missing file, mistakenly didn't add it in last commit

godmar commented on docs/subsystems/scoring/scoring.adoc:
```diff
@@ -0,0 +1,89 @@
+
+The scoring subsystem contains the xref:shooter.adoc[Shooter] and Indexer.
+
+It may be deemed beneficial that the scoring subsystem also includes the intake, to better coordinate intaking and scoring actions between the intake and the indexer-shooter combination.
```
Just reading this without background knowledge confuses me as to whether you're proposing a future change or describing the current state of affairs in this document.

aidnem replied:
I was just commenting on the current state of affairs. I had planned to have a discussion with @j0shuaS after he finishes indexer so we can decide on how to make them work together. Do you want me to elaborate more on this here?

godmar replied:
then I would word it as "We decided that it would be beneficial for the scoring subsystem to also include the intake, ...."

godmar commented on src/main/deploy/constants/config.json:
```diff
@@ -6,7 +6,12 @@
+        "CANConstants.json",
+        "RedFieldLocations.json",
+        "BlueFieldLocations.json",
+        "ShooterConstants.json",
```
Does that list `ShooterConstants.json` twice?

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterIOSim.java:
```diff
@@ -0,0 +1,71 @@
+
+    // Update sim
+    var motorVoltage = motorSimState.getMotorVoltage();
+    // TODO: Figure out why this method returns 0 on 1 of every 2 of the calls
```
Did you figure out this TODO?

aidnem replied:
Upon further testing, I realized it's because I forgot there were 2 ShooterIO objects. I was only running one at a time, but the other one was printing its "0.0" in the periodic loop too.

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterMechanism.java:
```diff
@@ -0,0 +1,572 @@
+
+    // The direction of the shot to take. This will be 90 degrees to the left of the robot's
+    // direction.
+    final Rotation2d shotDirection = robotPose.getRotation().plus(Rotation2d.kCCW_90deg);
```
You call both `Rotation2d` and `Translation2d` objects "*Direction".  I don't know if it would add clarity if you called it `shotAngle` or `shotRotation`.

aidnem replied:
Good catch. I added `Angle` and `Vector` to `shotDirection` and `rayDirection` to make it clearer what's what.

godmar commented on src/main/java/frc/robot/subsystems/scoring/shooter/ShooterMechanism.java:
```diff
@@ -0,0 +1,572 @@
+    boolean leftReady =
+        goalSpeeds.leftSpeed.isNear(
+            leftInputs.motorVelocity,
+            JsonConstants.shooterConstants.shooterVelocityEpsilonFraction);
```
Why is the permissive relative error expressed in terms of `shooterVelocityEpsilonFraction` independent of the shotDistance? Shouldn't the tolerance perhaps be tighter if you're farther away?  Then it could be expressed via an interval map, too.

aidnem replied:
Good point. For now, I think a flat RPM threshold will be enough because we can tune it to be small enough that it shouldn't cause us to miss. If it becomes a problem where our shots are very slow from close up because our tolerance is too close to make shots far away, or if we are missing shots from far away, I can always add another map to make it work during integration in a few minutes.

For now, I've swapped it to an RPM threshold.

godmar commented on src/main/java/frc/robot/util/PhoenixUtil.java:
```diff
@@ -17,11 +17,20 @@
     for (int i = 0; i < maxAttempts; i++) {
       var error = command.get();
-      if (error.isOK()) break;
+      if (error.isOK()) {
```
stray `;` plus `error.isOK()` sounds misnamed.

aidnem committed:
(#8) Rename rayDirection and shotDirection to rayDirectionVector and shotDirectionAngle for greater clarity

aidnem committed:
(#8) Swap shooter epsilon to a constant value rather than a fraction

aidnem committed:
(#8) Make PhoenixUtil status code name more descriptive; remove stray semicolon; format

aidnem committed:
(#8) Improve docs to reflect changes & decisions made during development

aidnem committed:
(#8) Add ScoringFeatureFlags and make shooter mechanisms Optional

aidnem committed:
(#8) Remove TODO to figure out 0.0 motor voltage; I figured it out; it was the second ShooterIO printing its idle voltage

aidnem committed:
(#8) Fix scoring.adoc typo: Interplating->Interpolating

avidraccoon commented on src/main/java/frc/robot/constants/subsystems/scoring/ShooterConstants.java:
```diff
@@ -1,5 +1,205 @@
+          EnvironmentHandler.getEnvironmentHandler().getEnvironmentPathProvider(),
+          new JSONSyncConfigBuilder().setPrettyPrinting(true).build());
+
+  public final String CANBusName = "canivore";
```
Is there a reason this should not be in CANConstants?

aidnem replied:
Good point. The original thought was that maybe superstructure would be on a different bus than drivebase but I think that's silly now.

avidraccoon commented on src/main/java/frc/robot/InitBindings.java:
```diff
@@ -0,0 +1,29 @@
+import frc.robot.subsystems.scoring.ScoringSubsystem;
+
+/** Methods to initialize bindings for each subsystems */
+public abstract class InitBindings {
```
Why is this abstract?

aidnem replied:
It's abstract so it can't be instantiated. I could also give it a private constructor if that's better style.

aidnem replied:
Made it final and gave it a private constructor

aidnem committed:
(#8) Make InitBindings final and give it private constructor

aidnem committed:
(#8) Move CAN bus name to CANConstants

godmar commented:
We could use instructions on how to test this in the simulator. 
We added an AutoChoice for closed-loop, set a target rpm, but the system never turned on the motors.

godmar commented on docs/subsystems/scoring/scoring.adoc:
```diff
@@ -0,0 +1,87 @@
+
+* Take the distance to that point and feed it into a map that uses InterpolatingDoubleTreeMap from WPILib to map distance to shooter speeds.
+
+** Potential challenge: Performance may be poor if we are constantly creating new ShooterSpeeds objects. ShooterSpeeds might be best converted into a class and then given a method to replace itself by interpolating between two values. Then, we could store one ShooterSpeeds object that gets replaced every in periodic with the current output of the map.
```
Just a sidenote. Unless you have actual prior experience (with excessive GC times) here, this smells a bit like premature optimization which as we know is the root of all evil. Modern GCs can efficiently deal with objects that die in the nursery; at worst, we may have to increase our nursery size.  I'd implement and switch to a mutable state model only if GC performance actually forced me to.

godmar commented on src/main/deploy/constants/comp/ShooterConstants.json:
```diff
@@ -0,0 +1,153 @@
+{
+  "CANBusName": "canivore",
+  "shooterVelocityEpsilon": {
+    "value": 60.0,
```
I believe you changed this to be to denote the absolute allowed deviation. So 60 rotations per second would be 3600 rotations per minute, or rpm.  That's half the speed of 7,200rpm hard drive. Is this correct as in can the actual shooter velocity really deviate by 3,600 rpm?

(Below it sets Shooter RPM to 100/50, but in rpm (?!))

aidnem replied:
Good catch! This was supposed to be one rotation per second.

godmar commented on src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java:
```diff
@@ -17,9 +20,9 @@ public class ScoringSubsystem extends MonitoredSubsystem {
-    this.indexer = indexer;
-    this.shooter = shooter;
+  private ScoringSubsystem(Optional<IndexerMechanism> indexer, Optional<ShooterMechanism> shooter) {
+    this.optionalIndexer = indexer;
```
I like the use of `Optional<T>` here, it may not even be necessary to prefix the instance field names with `optional` to keep things shorter.

aidnem replied:
I added `optional` to prevent cases like this:
```java
shooter.ifPresent(shooter -> shooter.testPeriodic()); // The name "shooter" means two different things on this line
```
However, if the convention is that this is okay, I'm willing to change it. I think that's cleaner, although slightly ambiguous. What's your opinion? I'm sure you've seen many more of these situations than I have.

godmar replied:
In my opinion, it's fine to keep the prefix off, but I'm not going to fight over it.
The difference between the optional shooter and the actual shooter is miniscule IMO.

aidnem replied:
👍Fixed it!

aidnem committed:
(#8) Fix 60 RPS shooter velocity epsilon -> 1 rps (60 rpm)

aidnem committed:
(#8) Fix bug where test modes were ignored due to shooter action enum and document shooter testing in sim

aidnem commented:
Turns out in my implementation of teleop, I accidentally made all test modes unusable (the shooter action would always override the current test mode command). Documentation has been added and the bug has been fixed.

aidnem committed:
(#8) Rename optional[Mechanism] to [mechanism] in shooter subsystem

godmar commented:
> Turns out in my implementation of teleop, I accidentally made all test modes unusable (the shooter action would always override the current test mode command). Documentation has been added and the bug has been fixed.

Weird. After trying for about an hour, we restarted vscode and all tools and all the sudden the right trigger worked in closed loop test mode. Before that, nothing.

But driving the robot still didn't work (or at least we couldn't figure out how to.)  Would it make sense to share/check in the config files for Glass/Elastic/AdvantageScope?  Also need to learn how to add the trajectory in odometry view.
